From e9cdcc7cb314586aec76b8b9979435f0af963443 Mon Sep 17 00:00:00 2001
From: Zhang Yu <zhangyub@uniontech.com>
Date: Tue, 15 Jun 2021 21:05:41 +0800
Subject: [PATCH] Fix recursion crash when calling setStyleSheet with `qproperty-styleSheet`

When calling `setStyleSheet` with property `qproperty-styleSheet`,
`QStyleSheetStyle::polish` will call`QStyleSheetStyle::setProperties`, and then`QStyleSheetStyle::setProperties` goes on to call `setProperty`.Because there is property `qproperty-styleSheet`, it will update stylesheet by calling QStyleSheetStyle::polish`.
This causes the recursive call to crash.

Fixes: QTBUG-94448
Pick-to: 5.15 6.2
Change-Id: I79c51192a939b0b62e5b1d0dcc90d38f79e28222
Reviewed-by: Jarek Kobus <jaroslaw.kobus@qt.io>
---

---
 src/widgets/styles/qstylesheetstyle.cpp                             |    3 +
 tests/auto/widgets/styles/qstylesheetstyle/tst_qstylesheetstyle.cpp |   18 ++++++++++
 2 files changed, 21 insertions(+)

--- a/src/widgets/styles/qstylesheetstyle.cpp
+++ b/src/widgets/styles/qstylesheetstyle.cpp
@@ -2626,6 +2626,9 @@ void QStyleSheetStyle::setProperties(QWi
         default: v = decl.d->values.at(0).variant; break;
         }
 
+        if (propertyL1 == QByteArrayView("styleSheet") && value == v)
+            continue;
+
         w->setProperty(propertyL1, v);
     }
 }
--- a/tests/auto/widgets/styles/qstylesheetstyle/tst_qstylesheetstyle.cpp
+++ b/tests/auto/widgets/styles/qstylesheetstyle/tst_qstylesheetstyle.cpp
@@ -94,6 +94,7 @@ private slots:
     void layoutSpacing();
 #endif
     void qproperty();
+    void qproperty_styleSheet();
     void palettePropagation_data();
     void palettePropagation();
     void fontPropagation_data();
@@ -677,6 +678,23 @@ void tst_QStyleSheetStyle::qproperty()
     QCOMPARE(pb.isChecked(), false);
 }
 
+void tst_QStyleSheetStyle::qproperty_styleSheet()
+{
+    QWidget w;
+    auto checkBox = new QCheckBox("check", &w);
+    QString sheet = R"(QCheckBox { qproperty-styleSheet: "QCheckBox { qproperty-text: foobar; }"; })";
+
+    QVERIFY(w.property("styleSheet").toString().isEmpty());
+
+    w.setStyleSheet(sheet);
+    QCOMPARE(checkBox->text(), "check");
+
+    //recursion crash
+    w.ensurePolished();
+    QCOMPARE(w.property("styleSheet").toString(), sheet);
+    QCOMPARE(checkBox->text(), "foobar");
+}
+
 namespace ns {
     class PushButton1 : public QPushButton {
         Q_OBJECT
